cloudinit/socket.py

```diff
@@ -0,0 +1,164 @@
+        raise OSError("Unsupported socket type")
+
+    with socket.socket(
+        socket.AF_UNIX, socket.SOCK_DGRAM | socket.SOCK_CLOEXEC
+    ) as sock:
+        LOG.info("Sending sd_notify(%s)", str(message))
+        sock.connect(socket_path)
+        sock.sendall(message.encode("ascii"))
+
+
+class SocketSync:
+    """A two way synchronization protocol over Unix domain sockets."""
+
+    def __init__(self, *names: str):
+        """Initialize a synchronization context.
+
+        1) Ensure that the socket directory exists.
+        2) Bind a socket for each stage.
+
+        Binding the sockets on initialization allows receipt of stage
+        "start" notifications prior to the cloud-init stage being ready to
+        start.
+
+        :param names: stage names, used as a unique identifiers
+        """
+        self.stage = ""
+        self.remote = ""
+        self.first_exception = ""
+        self.sockets = {
+            name: socket.socket(
+                socket.AF_UNIX, socket.SOCK_DGRAM | socket.SOCK_CLOEXEC
+            )
+            for name in names
+        }
+        # ensure the directory exists
+        os.makedirs(f"{DEFAULT_RUN_DIR}/share", mode=0o700, exist_ok=True)
+        # removing stale sockets and bind
+        for name, sock in self.sockets.items():
+            socket_path = f"{DEFAULT_RUN_DIR}/share/{name}.sock"
+            with suppress(FileNotFoundError):
+                os.remove(socket_path)
+            sock.bind(socket_path)
+
+    def __call__(self, stage: str):
+        """Set the stage before entering context.
+
+        This enables the context manager to be initialized separately from
+        each stage synchronization.
+
+        :param stage: the name of a stage to synchronize
+
+        Example:
+            sync = SocketSync("stage 1", "stage 2"):
+            with sync("stage 1"):
+                pass
+            with sync("stage 2"):
+                pass
+        """
+        if stage not in self.sockets:
+            raise ValueError(f"Invalid stage name: {stage}")
+        self.stage = stage
+        return self
+
+    def __enter__(self):
+        """Wait until a message has been received on this stage's socket.
+
+        Once the message has been received, enter the context.
+        """
+        if os.isatty(sys.stdin.fileno()):
+            LOG.info("Not syncing, stdin is a tty.")
+            return
+        sd_notify(
+            "STATUS=Waiting on external services to "
+            f"complete ({self.stage} stage)"
+        )
+        start_time = time.monotonic()
+        # block until init system sends us data
+        # the first value returned contains a message from the init system
+        #     (should be "start")
+        # the second value contains the path to a unix socket on which to
+        #     reply, which is expected to be /path/to/{self.stage}-return.sock
+        sock = self.sockets[self.stage]
+        chunk, self.remote = sock.recvfrom(5)
```

Comment:
Historically  systemd units have been kicked out of systemd boot goals due to misconfiguraiton ordering cycles (unit dependency chain issues) with other systemd units. Do we want to implement some sort of long timeout exception handling here to not block indefinitely and hard error that "cloud-init didn't start expected X stage in X mins" to unblock the rest of boot?

Reply:
> Historically systemd units have been kicked out of systemd boot goals due to misconfiguraiton ordering cycles (unit dependency chain issues) with other systemd units.
> Do we want to implement some sort of long timeout exception handling here to not block indefinitely and hard error that "cloud-init didn't start expected X stage in X mins" to unblock the rest of boot?

This is a condition which should be discovered before the image ever gets shipped, not something that should bite end users.

I really don't think that we want that. What you are referring to is a symptom seen at runtime of a problem that in almost all cases is introduced by a combination of packages that have been installed on the system (build time). Services in systemd can take an unknowable amount of time to complete and a timeout could break a real user of cloud-init that has to deal with long boot times. Such a timeout isn't a correct solution for this problem. The correct solution is to fix the ordering cycle. Adding such a timeout could break real users that have long boot times just to introduce a broken workaround for packagers.

Yes cloud-init will behave differently from before if booted out due to an ordering cycle. However I don't think that this is something that an end user should ever see unless they themselves are messing with the boot order or their OS vendor made a mistake - and even in those cases it still doesn't seem like a problem for cloud-init to attempt to solve at the expense of real users with slow boot times.

Reply:
> This is a condition which should be discovered before the image ever gets shipped, not something that should bite end users.

This does happen sometimes to (image creator) end-users who install supplemental systemd services in their golden images which has resulted in bogus bugs/questions about cloud-init not running right (because it was ejected from the boot goals due to systemd unit dependency conflicts).  I was wondering if there are ways we could make that condition a bit more conspicuous, but the symptoms below are probably "good enough" for those image creators to triage and determine what is causing the block in boot. 

Here are the symptoms a user would encounter if we don't instrument a timeout of some sort our failure mode in images where cloud-network.service were to be ejected from boot goals due to cyclic dependency ordering the system will be in the following state:
 - systemctl status` will indefinitely set at "starting" State.
 - systemctl list-jobs will not show the missing/disabled cloud-network.service but instead show cloud-final.service, cloud-init.target in waiting state and cloud-config.service sitting indefinitely in "running" state with not logs
 -  systemctl status cloud-init.service will show `Waiting on external services to complete (network stage)`
 - cloud-init status --format=yaml will show extended_status: running and no start times for init stage in the resulting YAML output.

These are probably reasonable enough breadcrumbs to look into on systems which exhibit this configuration problem and may not warrant further arbitrary timeouts which could adversely affect slow system startup.

Reply:
>> This is a condition which should be discovered before the image ever gets shipped, not something that should bite end users.
>
> This does happen sometimes to (image creator) end-users who install supplemental systemd services in their golden images which has resulted in bogus bugs/questions about cloud-init not running right (because it was ejected from the boot goals due to systemd unit dependency conflicts). I was wondering if there are ways we could make that condition a bit more conspicuous, but the symptoms below are probably "good enough" for those image creators to triage and determine what is causing the block in boot.

This can easily be checked with `systemd-analyze verify default.target` and checking the output, or by checking the journal for messages like: "Breaking ordering cycle by deleting job cloud-init.service".

I can add that to the docs in the [appropriate place](https://docs.cloud-init.io/en/latest/howto/debugging.html#cloud-init-never-finished-running) for anyone that runs into this. I'll make it a separate PR.